### DISCUSSION TITLE
CDT to generate makefiles with relative sources path from Linked Resources

### DISCUSSION THREAD

**Discussion by phatpaul** [eclipse-cdt/cdt#281] [Q&A]
Some of my source folders are "Linked Resources" with "Variable Relative Location" (i.e. PROJECT_LOC\..\..\src\app). But they still show full path in the generated makefiles on Windows (i.e. C_SRCS += C:/repos/fw/src/app/timer.c) and Linux (i.e. C_SRCS += /home/phatpaul/repos/fw/app/timer.c).

This is causing a headache with my team who use different OS, since we began to commit the makefiles to version control (for a scripted build and eventually CI).
How can I get CDT to write the makefiles with relative path for all files?

FYI, I found some posts asking for this in the past, but no good solution. 
https://stackoverflow.com/questions/22064413/linked-resource-absolute-paths-in-generated-makefiles-in-eclipse-cdt
https://stackoverflow.com/questions/33386873/configure-eclipse-cdt-to-generate-makefiles-with-relative-sources-paths
https://stackoverflow.com/questions/3997152/getting-eclipse-cdt-to-use-relative-include-paths-in-generated-makefiles (this one was specifically about "include paths" which is working, but source paths are still not working.)

EDIT: I'm using MCUXpresso IDE v11.7.0 which is based on Eclipse Version: 4.24.0.v20220607-0700 and Eclipse CDT version 10.7.0.202206081808

**Comment by jonahgraham** [comment]
I can't answer how to get relative paths in makefiles, the answers on StackOverflow you linked to are pretty accurate. You can ask your vendor to provide a patch (MCUXpresso) to fix this.

However, I can provide you an alternate solution that I know many people are using to automate their builds. Don't check in the makefiles, instead generate them in the scripted build using the headless builder. See https://gnu-mcu-eclipse.github.io/advanced/headless-builds/ from our related project which has a nice writeup.

**Comment by phatpaul** [comment]
Thanks Jonah for your reply and tips.  I don't usually get great support from NXP, so that's why I am reaching out to the opensource community.  It seems others are complaining about this on Eclipse platforms which are unrelated to MCUXpresso.

Is this the right project?  I assumed it is CDT that is responsible for generating the makefiles.  Would you please point me to the code that does this?

**Reply by jonahgraham** [reply]
> Thanks Jonah for your reply and tips. 

My pleasure.

> I don't usually get great support from NXP, so that's why I am reaching out to the opensource community. It seems others are complaining about this on Eclipse platforms which are unrelated to MCUXpresso.

Certainly NXP build on top of open source projects like CDT, unfortunately we don't seem them participating in the community much. If all their customers demanded it more then I think they would help out more, and perhaps contribute fixes that are important to their customers. Some of the engineers at NXP (e.g. Adrian) are occasionally involved, but it requires a decision from the top of the company to increase their involvement. As a result of all of that, I try to ask users of vendor products to contact their vendors.

> 
> Is this the right project? I assumed it is CDT that is responsible for generating the makefiles.

It is indeed.

> Would you please point me to the code that does this?

My best guess is this is the method to start your exploration at:

https://github.com/eclipse-cdt/cdt/blob/7c6bd5bdcb4cf8c31050aeeb102250a27a157728/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/gnu2/GnuMakefileGenerator.java#L2384

That method has this comment a bit further down:

https://github.com/eclipse-cdt/cdt/blob/7c6bd5bdcb4cf8c31050aeeb102250a27a157728/build/org.eclipse.cdt.managedbuilder.core/src/org/eclipse/cdt/managedbuilder/makegen/gnu2/GnuMakefileGenerator.java#L2492-L2494

Have a look at the [contribution guide](https://github.com/eclipse-cdt/cdt/blob/main/CONTRIBUTING.md) to get started, and don't hesitate to ask questions and we'll do the best we can to help support you providing a patch to fix this issue.
